numba/core/ir_utils.py

```diff
@@ -664,7 +664,12 @@ def remove_dead_block(block, lives, call_table, arg_aliases, alias_map,
         else:
             lives |= {v.name for v in stmt.list_vars()}
             if isinstance(stmt, ir.Assign):
-                lives.remove(lhs.name)
+                # make sure lhs is not used in rhs, e.g. a = g(a)
+                rhs_vars = set()
+                if isinstance(stmt.value, ir.Expr):
+                    rhs_vars = {v.name for v in stmt.value.list_vars()}
+                if lhs.name not in rhs_vars:
+                    lives.remove(lhs.name)
```

Comment:
```suggestion
                if isinstance(stmt.value, ir.Expr):
                    rhs_vars = {v.name for v in stmt.value.list_vars()}
                    if lhs.name not in rhs_vars:
                        lives.remove(lhs.name)
```
Would this work? The test condition exists solely if the assign value is an `Expr`?

Reply:
This needs `lives.remove(lhs.name)` for non-Expr nodes. Updated the code since previous structure wasn't clear..

Reply:
Would something like `a = print(a)` also potentially end up as a similar bug?

Reply:
The `Print` node doesn't have lhs so doesn't go through this. Also, `Print` has side effect and is never removed. 
```
    a = arg(0, name=a)                       ['a']
    $2load_global.0 = global(print: <built-in function print>) ['$2load_global.0']
    print(a)                                 ['a']
    a.1 = const(NoneType, None)              ['a.1']
    $12return_value.4 = cast(value=a.1)      ['$12return_value.4', 'a.1']
    return $12return_value.4                 ['$12return_value.4']
```

Reply:
Ah yes, thanks for checking.